pkg/controller/certificates/revisionmanager/revisionmanager_controller.go

```diff
@@ -0,0 +1,216 @@
+	client                   cmclient.Interface
+}
+
+type revision struct {
+	rev int
+	req *cmapi.CertificateRequest
+}
+
+func NewController(log logr.Logger, client cmclient.Interface, cmFactory cminformers.SharedInformerFactory) (*controller, workqueue.RateLimitingInterface, []cache.InformerSynced) {
+	// create a queue used to queue up items to be processed
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(time.Second*1, time.Second*30), ControllerName)
+
+	// obtain references to all the informers used by this controller
+	certificateInformer := cmFactory.Certmanager().V1().Certificates()
+	certificateRequestInformer := cmFactory.Certmanager().V1().CertificateRequests()
+
+	certificateInformer.Informer().AddEventHandler(&controllerpkg.QueuingEventHandler{Queue: queue})
+	certificateRequestInformer.Informer().AddEventHandler(&controllerpkg.BlockingEventHandler{
+		// Trigger reconciles on changes to any 'owned' CertificateRequest resources
+		WorkFunc: certificates.EnqueueCertificatesForResourceUsingPredicates(log, queue, certificateInformer.Lister(), labels.Everything(),
+			predicate.ResourceOwnerOf,
+		),
+	})
+
+	// build a list of InformerSynced functions that will be returned by the Register method.
+	// the controller will only begin processing items once all of these informers have synced.
+	mustSync := []cache.InformerSynced{
+		certificateRequestInformer.Informer().HasSynced,
+		certificateInformer.Informer().HasSynced,
+	}
+
+	return &controller{
+		certificateLister:        certificateInformer.Lister(),
+		certificateRequestLister: certificateRequestInformer.Lister(),
+		client:                   client,
+	}, queue, mustSync
+}
+
+// ProcessItem will attempt to garbage collect old CertificateRequests based
+// upon `spec.revisionHistoryLimit`. This controller will only act on
+// Certificates which are in a Ready state and this value is set.
+func (c *controller) ProcessItem(ctx context.Context, key string) error {
+	log := logf.FromContext(ctx).WithValues("key", key)
+
+	ctx = logf.NewContext(ctx, log)
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		log.Error(err, "invalid resource key passed to ProcessItem")
+		return nil
+	}
+
+	crt, err := c.certificateLister.Certificates(namespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		log.Error(err, "certificate not found for key")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	log = logf.WithResource(log, crt)
+
+	// If RevisionHistoryLimit is nil, don't attempt to garbage collect old
+	// CertificateRequests
+	if crt.Spec.RevisionHistoryLimit == nil {
+		return nil
+	}
+
+	limit := int(*crt.Spec.RevisionHistoryLimit)
+
+	// Only garbage collect over Certificates that are in a Ready=True condition.
+	if !apiutil.CertificateHasCondition(crt, cmapi.CertificateCondition{
+		Type:   cmapi.CertificateConditionReady,
+		Status: cmmeta.ConditionTrue,
+	}) {
+		return nil
+	}
```

Comment:
In the case where we have a string of Failed CRs and the Cert is still Failed, it would be prudent to expect the user would want to check the CR history and their failure reasons.